# Network Colocalization of Human and Rat BMI GWAS results
[Insert Description]

## Requirements
* `netcoloc` versions > . See installation instuctions at [url]
* `tqdm`
* `matplotlib-venn`
* `itertools`
* `other`


## Notebooks

**Main Notebooks:**  
* `1_Rat_Human_BMI_Network_Colocalization.ipynb`: ...
* `2_Rat_Human_BMI_Systems_Map.ipynb`: ...
* `3_Species_Specific_BMI_Networks.ipynb`: ...  

**Supplemental Notebooks:** 
* ....

## Function Files
* `analysis_functions.py` Functions outside of NetColoc used to generate the 
results in this study. 
* `plotting_functions.py` Functions used for plotting figures generated for this study.  
* `updated_netcoloc_functions.py` Contains modifications to the published NetColoc 
source code for use in this project. These modifications include:
 * ....

## Data files
* Seed genes:
 * `Data/inputs/`
* Pre-calculated propagation scores:
 * `Data/outputs/`
* Systems Map results:
 * `Data/outputs/`
* HCOP Human-Rat ortholog mapping:
 * `Data/Reference/`
  
## Other

**Mouse Genome Database (MGD) Files** These files can be automatically generated by the workflow 
but are included to maintain a consistent version of this database. These files can 
be updated by specifying `update=True` in the function `...`. 
* `MPheno_OBO.ontology`
* `MGI_PhenoGenoMP.rpt`
* `id2name_mp.txt`
* `id2namespace_mp.txt`
* `parsed_mp.txt`

**Environment File:**
`rat_genetics.yaml` provided to define environment used to develop the above code. 
