{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create table containing number of seed genes, loci per system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:51:42.285738Z",
     "start_time": "2022-07-18T19:51:37.858489Z"
    }
   },
   "outputs": [],
   "source": [
    "# import matplotlib\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# latex rendering of text in graphs\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "import community\n",
    "import ndex2\n",
    "from tqdm import tqdm\n",
    "\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "import mygene\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "sns.set(font_scale=1.5)\n",
    "\n",
    "sns.set_style('white')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:51:48.107366Z",
     "start_time": "2022-07-18T19:51:48.105093Z"
    }
   },
   "outputs": [],
   "source": [
    "cwd = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "DATADIR = os.path.join(cwd, \"Data/\")\n",
    "FIGDIR = os.path.join(cwd, \"Figures/rerun_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:51:48.842276Z",
     "start_time": "2022-07-18T19:51:48.578706Z"
    }
   },
   "outputs": [],
   "source": [
    "sys.path.append(cwd)\n",
    "from analysis_functions import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# load PCNet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:55:47.240639Z",
     "start_time": "2022-07-18T19:55:33.658316Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nodes:\n",
      "18820\n",
      "\n",
      "number of edges:\n",
      "2693109\n"
     ]
    }
   ],
   "source": [
    "pc_nodes, Gint = load_pcnet()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse mapped genes from human and rat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:55:47.423625Z",
     "start_time": "2022-07-18T19:55:47.353748Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1958\n"
     ]
    }
   ],
   "source": [
    "BMI_GIANT_pascal = pd.read_csv(DATADIR + 'inputs/GIANT_BMI_pascal.sum.genescores.txt', sep='\\t')\n",
    "BMI_GIANT_pascal.index=BMI_GIANT_pascal['gene_symbol']\n",
    "BMI_GIANT_pascal = BMI_GIANT_pascal.loc[list(np.intersect1d(BMI_GIANT_pascal.index.tolist(),pc_nodes))]\n",
    "bonf_p = .05/len(BMI_GIANT_pascal)\n",
    "\n",
    "h_bmi_genes = BMI_GIANT_pascal[BMI_GIANT_pascal['pvalue']<bonf_p].index.tolist()\n",
    "print(len(h_bmi_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:55:47.545697Z",
     "start_time": "2022-07-18T19:55:47.535943Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>numSnps</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33772366</td>\n",
       "      <td>33786699</td>\n",
       "      <td>-</td>\n",
       "      <td>127550</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>171</td>\n",
       "      <td>1.284589e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAT</th>\n",
       "      <td>chr4</td>\n",
       "      <td>170981372</td>\n",
       "      <td>171011372</td>\n",
       "      <td>-</td>\n",
       "      <td>51166</td>\n",
       "      <td>AADAT</td>\n",
       "      <td>269</td>\n",
       "      <td>7.579147e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS</th>\n",
       "      <td>chr16</td>\n",
       "      <td>70286296</td>\n",
       "      <td>70323412</td>\n",
       "      <td>-</td>\n",
       "      <td>16</td>\n",
       "      <td>AARS</td>\n",
       "      <td>31</td>\n",
       "      <td>1.125766e-13</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATK</th>\n",
       "      <td>chr17</td>\n",
       "      <td>79091095</td>\n",
       "      <td>79139872</td>\n",
       "      <td>-</td>\n",
       "      <td>9625</td>\n",
       "      <td>AATK</td>\n",
       "      <td>386</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>220074487</td>\n",
       "      <td>220085174</td>\n",
       "      <td>-</td>\n",
       "      <td>10058</td>\n",
       "      <td>ABCB6</td>\n",
       "      <td>112</td>\n",
       "      <td>4.647868e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome      start        end strand  gene_id gene_symbol  \\\n",
       "gene_symbol                                                                \n",
       "A3GALT2           chr1   33772366   33786699      -   127550     A3GALT2   \n",
       "AADAT             chr4  170981372  171011372      -    51166       AADAT   \n",
       "AARS             chr16   70286296   70323412      -       16        AARS   \n",
       "AATK             chr17   79091095   79139872      -     9625        AATK   \n",
       "ABCB6             chr2  220074487  220085174      -    10058       ABCB6   \n",
       "\n",
       "             numSnps        pvalue                                   Status  \n",
       "gene_symbol                                                                  \n",
       "A3GALT2          171  1.284589e-07                           DAVIES_SUCCESS  \n",
       "AADAT            269  7.579147e-07                           DAVIES_SUCCESS  \n",
       "AARS              31  1.125766e-13  DAVIES_LOWPRECISION_FAREBROTHER_SUCCESS  \n",
       "AATK             386  1.000000e-12     DAVIES_LOWPRECISION_FAREBROTHER_FAIL  \n",
       "ABCB6            112  4.647868e-07                           DAVIES_SUCCESS  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_bmi_seed_df = BMI_GIANT_pascal.loc[h_bmi_genes]\n",
    "human_bmi_seed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:51:55.015148Z",
     "start_time": "2022-07-18T19:51:54.999211Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "295\n"
     ]
    }
   ],
   "source": [
    "# set some parameters\n",
    "ratThresh='relaxed' # relaxed or stringent\n",
    "\n",
    "num_reps=1000\n",
    "\n",
    "# set type of gene mapper\n",
    "mapper = 'PASCAL' # can be PASCAL or PREDIXCAN\n",
    "\n",
    "adj_type = 'bonf' # can be bonf or BH\n",
    "# bonf_p = .05/len(BMI_GIANT_pascal) #0.25/len(BMI_GIANT_pascal)\n",
    "# print(bonf_p)\n",
    "\n",
    "excl_rat_seeds = False # if true, exclude rat seed genes from relevant phenotype\n",
    "\n",
    "# read in rat BMI seed genes, corresponding to parameters above\n",
    "rat_bmi_genes=pd.read_csv(DATADIR + 'inputs/ratBMI_seed_'+ratThresh+'.txt',sep='\\t')['0'].tolist()\n",
    "print(len(rat_bmi_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:12:02.625309Z",
     "start_time": "2022-07-18T20:12:02.557429Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start-End</th>\n",
       "      <th>nSNPS</th>\n",
       "      <th>TopSNP P-Value</th>\n",
       "      <th>TopSNP Position</th>\n",
       "      <th>Binarized Heat</th>\n",
       "      <th>Negative Log</th>\n",
       "      <th>Size Adjusted Heat</th>\n",
       "      <th>HumanGene</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HumanGene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRLHR</th>\n",
       "      <td>Prlhr</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281754471-281756159</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.310000e-12</td>\n",
       "      <td>281756885.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.136172</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>PRLHR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM204A</th>\n",
       "      <td>Fam204a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281343691-281390632\\t281343691-281403699\\t2813...</td>\n",
       "      <td>141.0</td>\n",
       "      <td>3.470000e-11</td>\n",
       "      <td>281411373.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.459144</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>FAM204A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>LOC108349713</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281397475-281406785</td>\n",
       "      <td>62.0</td>\n",
       "      <td>3.470000e-11</td>\n",
       "      <td>281411373.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.459144</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>LOC102556108</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281289719-281301391\\t281299202-281301391</td>\n",
       "      <td>33.0</td>\n",
       "      <td>3.700000e-11</td>\n",
       "      <td>281303363.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.431224</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>LOC102556164</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281227922-281318201\\t281227922-281318201\\t2812...</td>\n",
       "      <td>186.0</td>\n",
       "      <td>3.700000e-11</td>\n",
       "      <td>281303363.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.431224</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene Chrom  \\\n",
       "HumanGene                       \n",
       "PRLHR             Prlhr  chr1   \n",
       "FAM204A         Fam204a  chr1   \n",
       "NaN        LOC108349713  chr1   \n",
       "NaN        LOC102556108  chr1   \n",
       "NaN        LOC102556164  chr1   \n",
       "\n",
       "                                                   Start-End  nSNPS  \\\n",
       "HumanGene                                                             \n",
       "PRLHR                                    281754471-281756159   23.0   \n",
       "FAM204A    281343691-281390632\\t281343691-281403699\\t2813...  141.0   \n",
       "NaN                                      281397475-281406785   62.0   \n",
       "NaN                 281289719-281301391\\t281299202-281301391   33.0   \n",
       "NaN        281227922-281318201\\t281227922-281318201\\t2812...  186.0   \n",
       "\n",
       "           TopSNP P-Value  TopSNP Position  Binarized Heat  Negative Log  \\\n",
       "HumanGene                                                                  \n",
       "PRLHR        7.310000e-12      281756885.0             1.0     11.136172   \n",
       "FAM204A      3.470000e-11      281411373.0             1.0     10.459144   \n",
       "NaN          3.470000e-11      281411373.0             1.0     10.459144   \n",
       "NaN          3.700000e-11      281303363.0             1.0     10.431224   \n",
       "NaN          3.700000e-11      281303363.0             1.0     10.431224   \n",
       "\n",
       "           Size Adjusted Heat HumanGene  Unnamed: 11  \n",
       "HumanGene                                             \n",
       "PRLHR                4.861444     PRLHR          NaN  \n",
       "FAM204A              4.861444   FAM204A          NaN  \n",
       "NaN                  4.861444       NaN          NaN  \n",
       "NaN                  4.861444       NaN          NaN  \n",
       "NaN                  4.861444       NaN          NaN  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_bmi_df = pd.read_csv(DATADIR + \"inputs/rat_BMI_summary_stats_wo_tail.tsv\", index_col=0,sep=\"\\t\") \n",
    "rat_bmi_df.index=rat_bmi_df['HumanGene']\n",
    "rat_bmi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:51:56.625978Z",
     "start_time": "2022-07-18T19:51:56.608158Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "295\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start-End</th>\n",
       "      <th>nSNPS</th>\n",
       "      <th>TopSNP P-Value</th>\n",
       "      <th>TopSNP Position</th>\n",
       "      <th>Binarized Heat</th>\n",
       "      <th>Negative Log</th>\n",
       "      <th>Size Adjusted Heat</th>\n",
       "      <th>HumanGene</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HumanGene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRLHR</th>\n",
       "      <td>Prlhr</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281754471-281756159</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.310000e-12</td>\n",
       "      <td>281756885.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.136172</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>PRLHR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM204A</th>\n",
       "      <td>Fam204a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281343691-281390632,281343691-281403699,281347...</td>\n",
       "      <td>141.0</td>\n",
       "      <td>3.470000e-11</td>\n",
       "      <td>281411373.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.459144</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>FAM204A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB11FIP2</th>\n",
       "      <td>Rab11fip2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281065345-281101163</td>\n",
       "      <td>141.0</td>\n",
       "      <td>1.700000e-10</td>\n",
       "      <td>281109613.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.770121</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>RAB11FIP2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACUL1</th>\n",
       "      <td>Cacul1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281814225-281874675,281816405-281874511</td>\n",
       "      <td>84.0</td>\n",
       "      <td>2.250000e-10</td>\n",
       "      <td>281811420.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.647775</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>CACUL1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARSG</th>\n",
       "      <td>Arsg</td>\n",
       "      <td>chr10</td>\n",
       "      <td>97722549-97859975,97733458-97859975,97733459-9...</td>\n",
       "      <td>355.0</td>\n",
       "      <td>7.620000e-10</td>\n",
       "      <td>97797154.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.117885</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>ARSG</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKIA</th>\n",
       "      <td>Pkia</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96593184-96608706,96593184-96608708,96593184-9...</td>\n",
       "      <td>283.0</td>\n",
       "      <td>9.330000e-05</td>\n",
       "      <td>96656161.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.029900</td>\n",
       "      <td>3.919947</td>\n",
       "      <td>PKIA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GK5</th>\n",
       "      <td>Gk5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>103929092-104001916</td>\n",
       "      <td>59.0</td>\n",
       "      <td>9.460000e-05</td>\n",
       "      <td>103969544.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.024207</td>\n",
       "      <td>3.916477</td>\n",
       "      <td>GK5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MRC2</th>\n",
       "      <td>Mrc2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>93520271-93581599,93520435-93580474</td>\n",
       "      <td>259.0</td>\n",
       "      <td>9.680000e-05</td>\n",
       "      <td>93546093.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.013937</td>\n",
       "      <td>3.917748</td>\n",
       "      <td>MRC2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABR</th>\n",
       "      <td>Abr</td>\n",
       "      <td>chr10</td>\n",
       "      <td>64565120-64657079</td>\n",
       "      <td>48.0</td>\n",
       "      <td>9.830000e-05</td>\n",
       "      <td>64611901.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.007652</td>\n",
       "      <td>3.880712</td>\n",
       "      <td>ABR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENDOV</th>\n",
       "      <td>Endov</td>\n",
       "      <td>chr10</td>\n",
       "      <td>108630722-108646679,108630793-108646679,108630...</td>\n",
       "      <td>53.0</td>\n",
       "      <td>9.870000e-05</td>\n",
       "      <td>108629743.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.005507</td>\n",
       "      <td>3.954869</td>\n",
       "      <td>ENDOV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>295 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene  Chrom  \\\n",
       "HumanGene                     \n",
       "PRLHR          Prlhr   chr1   \n",
       "FAM204A      Fam204a   chr1   \n",
       "RAB11FIP2  Rab11fip2   chr1   \n",
       "CACUL1        Cacul1   chr1   \n",
       "ARSG            Arsg  chr10   \n",
       "...              ...    ...   \n",
       "PKIA            Pkia   chr2   \n",
       "GK5              Gk5   chr8   \n",
       "MRC2            Mrc2  chr10   \n",
       "ABR              Abr  chr10   \n",
       "ENDOV          Endov  chr10   \n",
       "\n",
       "                                                   Start-End  nSNPS  \\\n",
       "HumanGene                                                             \n",
       "PRLHR                                    281754471-281756159   23.0   \n",
       "FAM204A    281343691-281390632,281343691-281403699,281347...  141.0   \n",
       "RAB11FIP2                                281065345-281101163  141.0   \n",
       "CACUL1               281814225-281874675,281816405-281874511   84.0   \n",
       "ARSG       97722549-97859975,97733458-97859975,97733459-9...  355.0   \n",
       "...                                                      ...    ...   \n",
       "PKIA       96593184-96608706,96593184-96608708,96593184-9...  283.0   \n",
       "GK5                                      103929092-104001916   59.0   \n",
       "MRC2                     93520271-93581599,93520435-93580474  259.0   \n",
       "ABR                                        64565120-64657079   48.0   \n",
       "ENDOV      108630722-108646679,108630793-108646679,108630...   53.0   \n",
       "\n",
       "           TopSNP P-Value  TopSNP Position  Binarized Heat  Negative Log  \\\n",
       "HumanGene                                                                  \n",
       "PRLHR        7.310000e-12      281756885.0             1.0     11.136172   \n",
       "FAM204A      3.470000e-11      281411373.0             1.0     10.459144   \n",
       "RAB11FIP2    1.700000e-10      281109613.0             1.0      9.770121   \n",
       "CACUL1       2.250000e-10      281811420.0             1.0      9.647775   \n",
       "ARSG         7.620000e-10       97797154.0             1.0      9.117885   \n",
       "...                   ...              ...             ...           ...   \n",
       "PKIA         9.330000e-05       96656161.0             0.0      4.029900   \n",
       "GK5          9.460000e-05      103969544.0             0.0      4.024207   \n",
       "MRC2         9.680000e-05       93546093.0             0.0      4.013937   \n",
       "ABR          9.830000e-05       64611901.0             0.0      4.007652   \n",
       "ENDOV        9.870000e-05      108629743.0             0.0      4.005507   \n",
       "\n",
       "           Size Adjusted Heat  HumanGene  Unnamed: 11  \n",
       "HumanGene                                              \n",
       "PRLHR                4.861444      PRLHR          NaN  \n",
       "FAM204A              4.861444    FAM204A          NaN  \n",
       "RAB11FIP2            4.861444  RAB11FIP2          NaN  \n",
       "CACUL1               4.861444     CACUL1          NaN  \n",
       "ARSG                 4.861444       ARSG          NaN  \n",
       "...                       ...        ...          ...  \n",
       "PKIA                 3.919947       PKIA          NaN  \n",
       "GK5                  3.916477        GK5          NaN  \n",
       "MRC2                 3.917748       MRC2          NaN  \n",
       "ABR                  3.880712        ABR          NaN  \n",
       "ENDOV                3.954869      ENDOV          NaN  \n",
       "\n",
       "[295 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_bmi_seed_df = rat_bmi_df.loc[rat_bmi_genes]\n",
    "print(len(rat_bmi_seed_df))\n",
    "rat_bmi_seed_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the clustering results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:00:22.600489Z",
     "start_time": "2022-07-18T20:00:22.588519Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD_MemberList</th>\n",
       "      <th>frac_d1_seeds</th>\n",
       "      <th>frac_d2_seeds</th>\n",
       "      <th>CD_MemberList_Size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C877</th>\n",
       "      <td>HELZ EMX2 HIST1H3A RAD51 SYT12 GYPC KDM3B ERC1...</td>\n",
       "      <td>0.353583</td>\n",
       "      <td>0.177570</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C898</th>\n",
       "      <td>STX4 LSAMP SPOCK1 SGCB STXBP5 TRP-AGG2-6 GPRC5...</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C904</th>\n",
       "      <td>BCL11B PKIB PKIA LMO1 NR2F1 LDB2 ZFPM2 GATA3 A...</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C906</th>\n",
       "      <td>PROM2 SPINT1 TNIK PKP3 SSH3 FEZ2 CORIN WWC1 CAPN1</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C911</th>\n",
       "      <td>ZNF629 ZKSCAN5 LINGO1 ZNF202 ZKSCAN2 ZNF263 ZN...</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CD_MemberList  frac_d1_seeds  \\\n",
       "C877  HELZ EMX2 HIST1H3A RAD51 SYT12 GYPC KDM3B ERC1...       0.353583   \n",
       "C898  STX4 LSAMP SPOCK1 SGCB STXBP5 TRP-AGG2-6 GPRC5...       0.363636   \n",
       "C904  BCL11B PKIB PKIA LMO1 NR2F1 LDB2 ZFPM2 GATA3 A...       0.300000   \n",
       "C906  PROM2 SPINT1 TNIK PKP3 SSH3 FEZ2 CORIN WWC1 CAPN1       0.333333   \n",
       "C911  ZNF629 ZKSCAN5 LINGO1 ZNF202 ZKSCAN2 ZNF263 ZN...       0.571429   \n",
       "\n",
       "      frac_d2_seeds  CD_MemberList_Size  \n",
       "C877       0.177570                 642  \n",
       "C898       0.181818                  11  \n",
       "C904       0.200000                  10  \n",
       "C906       0.111111                   9  \n",
       "C911       0.142857                   7  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the clustering results\n",
    "clust_df = pd.read_csv(DATADIR + \"outputs/BMI_hierarchy_data.tsv\", sep=\"\\t\", index_col=0)\n",
    "clust_df[\"CD_MemberList_Size\"] = clust_df.CD_MemberList.apply(lambda x: len(x.split(\" \")))\n",
    "print(len(clust_df))\n",
    "clust_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ID which gene pairs are in the same genomic loci (within ~ 250KB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:52:00.867874Z",
     "start_time": "2022-07-18T19:52:00.854315Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Start-End</th>\n",
       "      <th>nSNPS</th>\n",
       "      <th>TopSNP P-Value</th>\n",
       "      <th>TopSNP Position</th>\n",
       "      <th>Binarized Heat</th>\n",
       "      <th>Negative Log</th>\n",
       "      <th>Size Adjusted Heat</th>\n",
       "      <th>HumanGene</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HumanGene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRLHR</th>\n",
       "      <td>Prlhr</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281754471-281756159</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.310000e-12</td>\n",
       "      <td>281756885.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.136172</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>PRLHR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM204A</th>\n",
       "      <td>Fam204a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281343691-281390632,281343691-281403699,281347...</td>\n",
       "      <td>141.0</td>\n",
       "      <td>3.470000e-11</td>\n",
       "      <td>281411373.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.459144</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>FAM204A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB11FIP2</th>\n",
       "      <td>Rab11fip2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281065345-281101163</td>\n",
       "      <td>141.0</td>\n",
       "      <td>1.700000e-10</td>\n",
       "      <td>281109613.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.770121</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>RAB11FIP2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACUL1</th>\n",
       "      <td>Cacul1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>281814225-281874675,281816405-281874511</td>\n",
       "      <td>84.0</td>\n",
       "      <td>2.250000e-10</td>\n",
       "      <td>281811420.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.647775</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>CACUL1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARSG</th>\n",
       "      <td>Arsg</td>\n",
       "      <td>chr10</td>\n",
       "      <td>97722549-97859975,97733458-97859975,97733459-9...</td>\n",
       "      <td>355.0</td>\n",
       "      <td>7.620000e-10</td>\n",
       "      <td>97797154.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.117885</td>\n",
       "      <td>4.861444</td>\n",
       "      <td>ARSG</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKIA</th>\n",
       "      <td>Pkia</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96593184-96608706,96593184-96608708,96593184-9...</td>\n",
       "      <td>283.0</td>\n",
       "      <td>9.330000e-05</td>\n",
       "      <td>96656161.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.029900</td>\n",
       "      <td>3.919947</td>\n",
       "      <td>PKIA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GK5</th>\n",
       "      <td>Gk5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>103929092-104001916</td>\n",
       "      <td>59.0</td>\n",
       "      <td>9.460000e-05</td>\n",
       "      <td>103969544.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.024207</td>\n",
       "      <td>3.916477</td>\n",
       "      <td>GK5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MRC2</th>\n",
       "      <td>Mrc2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>93520271-93581599,93520435-93580474</td>\n",
       "      <td>259.0</td>\n",
       "      <td>9.680000e-05</td>\n",
       "      <td>93546093.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.013937</td>\n",
       "      <td>3.917748</td>\n",
       "      <td>MRC2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABR</th>\n",
       "      <td>Abr</td>\n",
       "      <td>chr10</td>\n",
       "      <td>64565120-64657079</td>\n",
       "      <td>48.0</td>\n",
       "      <td>9.830000e-05</td>\n",
       "      <td>64611901.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.007652</td>\n",
       "      <td>3.880712</td>\n",
       "      <td>ABR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENDOV</th>\n",
       "      <td>Endov</td>\n",
       "      <td>chr10</td>\n",
       "      <td>108630722-108646679,108630793-108646679,108630...</td>\n",
       "      <td>53.0</td>\n",
       "      <td>9.870000e-05</td>\n",
       "      <td>108629743.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.005507</td>\n",
       "      <td>3.954869</td>\n",
       "      <td>ENDOV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>295 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene  Chrom  \\\n",
       "HumanGene                     \n",
       "PRLHR          Prlhr   chr1   \n",
       "FAM204A      Fam204a   chr1   \n",
       "RAB11FIP2  Rab11fip2   chr1   \n",
       "CACUL1        Cacul1   chr1   \n",
       "ARSG            Arsg  chr10   \n",
       "...              ...    ...   \n",
       "PKIA            Pkia   chr2   \n",
       "GK5              Gk5   chr8   \n",
       "MRC2            Mrc2  chr10   \n",
       "ABR              Abr  chr10   \n",
       "ENDOV          Endov  chr10   \n",
       "\n",
       "                                                   Start-End  nSNPS  \\\n",
       "HumanGene                                                             \n",
       "PRLHR                                    281754471-281756159   23.0   \n",
       "FAM204A    281343691-281390632,281343691-281403699,281347...  141.0   \n",
       "RAB11FIP2                                281065345-281101163  141.0   \n",
       "CACUL1               281814225-281874675,281816405-281874511   84.0   \n",
       "ARSG       97722549-97859975,97733458-97859975,97733459-9...  355.0   \n",
       "...                                                      ...    ...   \n",
       "PKIA       96593184-96608706,96593184-96608708,96593184-9...  283.0   \n",
       "GK5                                      103929092-104001916   59.0   \n",
       "MRC2                     93520271-93581599,93520435-93580474  259.0   \n",
       "ABR                                        64565120-64657079   48.0   \n",
       "ENDOV      108630722-108646679,108630793-108646679,108630...   53.0   \n",
       "\n",
       "           TopSNP P-Value  TopSNP Position  Binarized Heat  Negative Log  \\\n",
       "HumanGene                                                                  \n",
       "PRLHR        7.310000e-12      281756885.0             1.0     11.136172   \n",
       "FAM204A      3.470000e-11      281411373.0             1.0     10.459144   \n",
       "RAB11FIP2    1.700000e-10      281109613.0             1.0      9.770121   \n",
       "CACUL1       2.250000e-10      281811420.0             1.0      9.647775   \n",
       "ARSG         7.620000e-10       97797154.0             1.0      9.117885   \n",
       "...                   ...              ...             ...           ...   \n",
       "PKIA         9.330000e-05       96656161.0             0.0      4.029900   \n",
       "GK5          9.460000e-05      103969544.0             0.0      4.024207   \n",
       "MRC2         9.680000e-05       93546093.0             0.0      4.013937   \n",
       "ABR          9.830000e-05       64611901.0             0.0      4.007652   \n",
       "ENDOV        9.870000e-05      108629743.0             0.0      4.005507   \n",
       "\n",
       "           Size Adjusted Heat  HumanGene  Unnamed: 11  \n",
       "HumanGene                                              \n",
       "PRLHR                4.861444      PRLHR          NaN  \n",
       "FAM204A              4.861444    FAM204A          NaN  \n",
       "RAB11FIP2            4.861444  RAB11FIP2          NaN  \n",
       "CACUL1               4.861444     CACUL1          NaN  \n",
       "ARSG                 4.861444       ARSG          NaN  \n",
       "...                       ...        ...          ...  \n",
       "PKIA                 3.919947       PKIA          NaN  \n",
       "GK5                  3.916477        GK5          NaN  \n",
       "MRC2                 3.917748       MRC2          NaN  \n",
       "ABR                  3.880712        ABR          NaN  \n",
       "ENDOV                3.954869      ENDOV          NaN  \n",
       "\n",
       "[295 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_bmi_seed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:55:05.939584Z",
     "start_time": "2022-07-18T19:55:03.747956Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C877\n",
      "C898\n",
      "C904\n",
      "C906\n",
      "C911\n",
      "C913\n",
      "C914\n",
      "C915\n",
      "C924\n",
      "C925\n",
      "C926\n",
      "C930\n",
      "C931\n",
      "C933\n",
      "C935\n",
      "C937\n",
      "C878\n",
      "C880\n",
      "C882\n",
      "C885\n",
      "C887\n",
      "C888\n",
      "C889\n",
      "C890\n",
      "C892\n",
      "C895\n",
      "C883\n",
      "C879\n",
      "C932\n",
      "C901\n",
      "C902\n",
      "C907\n",
      "C909\n",
      "C884\n",
      "C893\n",
      "C894\n",
      "C896\n",
      "C929\n",
      "C899\n",
      "C934\n",
      "C903\n",
      "C881\n",
      "C916\n",
      "C919\n",
      "C921\n",
      "C920\n",
      "C905\n",
      "C922\n",
      "C910\n",
      "C923\n",
      "C886\n",
      "C912\n",
      "C897\n",
      "C891\n",
      "C908\n",
      "C900\n",
      "C928\n",
      "C917\n",
      "C918\n",
      "C927\n",
      "C936\n"
     ]
    }
   ],
   "source": [
    "num_rat_loci=[]\n",
    "num_rat_seeds=[]\n",
    "for focal_clust in clust_df.index.tolist():\n",
    "    print(focal_clust)\n",
    "\n",
    "    focal_genes = clust_df['CD_MemberList'].loc[focal_clust].split(' ')\n",
    "\n",
    "    if len(np.intersect1d(focal_genes,rat_bmi_genes))>0: # only compute loci if there are any seed genes in the cluster\n",
    "        rat_bmi_seed_focal=rat_bmi_seed_df[rat_bmi_seed_df.index.isin(focal_genes)].dropna(subset=['HumanGene'])\n",
    "        num_rat_seeds.append(len(rat_bmi_seed_focal))\n",
    "\n",
    "        # how many seed genes are within 250 KB of each other?\n",
    "        start_list = []\n",
    "        end_list = []\n",
    "        chr_list = []\n",
    "        for g in rat_bmi_seed_focal.index.tolist():\n",
    "            # take the min start and max end\n",
    "            start_end_temp = rat_bmi_seed_focal.loc[g]['Start-End'].split(',')\n",
    "            start_temp = [int(s.split('-')[0]) for s in start_end_temp]\n",
    "            end_temp = [int(s.split('-')[1]) for s in start_end_temp]\n",
    "            start_temp=min(start_temp)\n",
    "            end_temp = max(end_temp)\n",
    "\n",
    "            chr_temp = rat_bmi_seed_focal.loc[g]['Chrom']\n",
    "\n",
    "            start_list.append(start_temp)\n",
    "            end_list.append(end_temp)\n",
    "            chr_list.append(chr_temp)\n",
    "\n",
    "        df_temp = pd.DataFrame({'start':start_list,'end':end_list,'chr':chr_list},index=rat_bmi_seed_focal.index.tolist())\n",
    "\n",
    "\n",
    "        # loop over all gene pairs, compute distance between start and end, when on same chromosome\n",
    "        g1list = []\n",
    "        g2list = []\n",
    "        dist_list = []\n",
    "        chrom1list = []\n",
    "        chrom2list = []\n",
    "\n",
    "        genes_temp = df_temp.index.tolist()\n",
    "        for i in np.arange(len(genes_temp)-1):\n",
    "            g1 = genes_temp[i]\n",
    "            chr1 = df_temp.loc[g1]['chr']\n",
    "            start1 = df_temp.loc[g1]['start']\n",
    "            end1 = df_temp.loc[g1]['end']\n",
    "            for j in np.arange(i+1,len(genes_temp)): \n",
    "                g2 = genes_temp[j]\n",
    "                chr2 = df_temp.loc[g2]['chr']\n",
    "                start2 = df_temp.loc[g2]['start']\n",
    "                end2 = df_temp.loc[g2]['end']\n",
    "\n",
    "                if chr1==chr2:\n",
    "                    # check if they overlap\n",
    "                    if (start2<end1)&(start1<start2):\n",
    "                        dist_list.append(0)\n",
    "                    elif (start1<end2)&(start2<start1):\n",
    "                        dist_list.append(0)\n",
    "                    elif (start2-end1)>0: # check which is ahead\n",
    "                        dist_list.append(start2-end1)\n",
    "                    else:\n",
    "                        dist_list.append(start1-end2)\n",
    "                else:\n",
    "                    dist_list.append(np.nan) # nans if they are diff chrom\n",
    "\n",
    "                g1list.append(g1)\n",
    "                g2list.append(g2)\n",
    "                chrom1list.append(chr1)\n",
    "                chrom2list.append(chr2)\n",
    "\n",
    "        dist_df_temp = pd.DataFrame({'g1':g1list,'g2':g2list,\n",
    "                                     'chr1':chrom1list,'chr2':chrom2list,\n",
    "                                     'dist':dist_list})\n",
    "\n",
    "\n",
    "\n",
    "        dist_df_temp['within_250KB']=dist_df_temp['dist']<250000\n",
    "        elist_temp = zip(dist_df_temp[dist_df_temp['within_250KB']==True]['g1'],dist_df_temp[dist_df_temp['within_250KB']==True]['g2'])\n",
    "\n",
    "        # make a network from this\n",
    "        G_dist_temp = nx.from_edgelist(elist_temp)\n",
    "        # number communities + number genes not in a shared locus ==> total number of loci\n",
    "        num_rat_loci.append(len(pd.Series(community.best_partition(G_dist_temp), dtype=\"object\").value_counts())+len(np.setdiff1d(genes_temp,G_dist_temp.nodes())))\n",
    "    else:\n",
    "        num_rat_loci.append(0)\n",
    "        num_rat_seeds.append(0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:56:43.631944Z",
     "start_time": "2022-07-18T19:56:43.625105Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>numSnps</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>33772366</td>\n",
       "      <td>33786699</td>\n",
       "      <td>-</td>\n",
       "      <td>127550</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>171</td>\n",
       "      <td>1.284589e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAT</th>\n",
       "      <td>chr4</td>\n",
       "      <td>170981372</td>\n",
       "      <td>171011372</td>\n",
       "      <td>-</td>\n",
       "      <td>51166</td>\n",
       "      <td>AADAT</td>\n",
       "      <td>269</td>\n",
       "      <td>7.579147e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS</th>\n",
       "      <td>chr16</td>\n",
       "      <td>70286296</td>\n",
       "      <td>70323412</td>\n",
       "      <td>-</td>\n",
       "      <td>16</td>\n",
       "      <td>AARS</td>\n",
       "      <td>31</td>\n",
       "      <td>1.125766e-13</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_SUCCESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATK</th>\n",
       "      <td>chr17</td>\n",
       "      <td>79091095</td>\n",
       "      <td>79139872</td>\n",
       "      <td>-</td>\n",
       "      <td>9625</td>\n",
       "      <td>AATK</td>\n",
       "      <td>386</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>DAVIES_LOWPRECISION_FAREBROTHER_FAIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>220074487</td>\n",
       "      <td>220085174</td>\n",
       "      <td>-</td>\n",
       "      <td>10058</td>\n",
       "      <td>ABCB6</td>\n",
       "      <td>112</td>\n",
       "      <td>4.647868e-07</td>\n",
       "      <td>DAVIES_SUCCESS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome      start        end strand  gene_id gene_symbol  \\\n",
       "gene_symbol                                                                \n",
       "A3GALT2           chr1   33772366   33786699      -   127550     A3GALT2   \n",
       "AADAT             chr4  170981372  171011372      -    51166       AADAT   \n",
       "AARS             chr16   70286296   70323412      -       16        AARS   \n",
       "AATK             chr17   79091095   79139872      -     9625        AATK   \n",
       "ABCB6             chr2  220074487  220085174      -    10058       ABCB6   \n",
       "\n",
       "             numSnps        pvalue                                   Status  \n",
       "gene_symbol                                                                  \n",
       "A3GALT2          171  1.284589e-07                           DAVIES_SUCCESS  \n",
       "AADAT            269  7.579147e-07                           DAVIES_SUCCESS  \n",
       "AARS              31  1.125766e-13  DAVIES_LOWPRECISION_FAREBROTHER_SUCCESS  \n",
       "AATK             386  1.000000e-12     DAVIES_LOWPRECISION_FAREBROTHER_FAIL  \n",
       "ABCB6            112  4.647868e-07                           DAVIES_SUCCESS  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_bmi_seed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:57:22.879666Z",
     "start_time": "2022-07-18T19:57:22.866447Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>numSnps</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AATK</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AATK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABCB8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADORA2B</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADORA2B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZKSCAN5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZKSCAN5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF142</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF629</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF768</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZSWIM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>227 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chromosome  start  end strand  gene_id gene_symbol  numSnps  \\\n",
       "gene_symbol                                                               \n",
       "AATK               NaN    NaN  NaN    NaN      NaN        AATK      NaN   \n",
       "ABCB8              NaN    NaN  NaN    NaN      NaN       ABCB8      NaN   \n",
       "ABL1               NaN    NaN  NaN    NaN      NaN        ABL1      NaN   \n",
       "ADORA2B            NaN    NaN  NaN    NaN      NaN     ADORA2B      NaN   \n",
       "AGAP1              NaN    NaN  NaN    NaN      NaN       AGAP1      NaN   \n",
       "...                ...    ...  ...    ...      ...         ...      ...   \n",
       "ZKSCAN5            NaN    NaN  NaN    NaN      NaN     ZKSCAN5      NaN   \n",
       "ZNF142             NaN    NaN  NaN    NaN      NaN      ZNF142      NaN   \n",
       "ZNF629             NaN    NaN  NaN    NaN      NaN      ZNF629      NaN   \n",
       "ZNF768             NaN    NaN  NaN    NaN      NaN      ZNF768      NaN   \n",
       "ZSWIM6             NaN    NaN  NaN    NaN      NaN      ZSWIM6      NaN   \n",
       "\n",
       "             pvalue Status  \n",
       "gene_symbol                 \n",
       "AATK            NaN    NaN  \n",
       "ABCB8           NaN    NaN  \n",
       "ABL1            NaN    NaN  \n",
       "ADORA2B         NaN    NaN  \n",
       "AGAP1           NaN    NaN  \n",
       "...             ...    ...  \n",
       "ZKSCAN5         NaN    NaN  \n",
       "ZNF142          NaN    NaN  \n",
       "ZNF629          NaN    NaN  \n",
       "ZNF768          NaN    NaN  \n",
       "ZSWIM6          NaN    NaN  \n",
       "\n",
       "[227 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_bmi_seed_focal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:58:31.769969Z",
     "start_time": "2022-07-18T19:58:23.539698Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C877\n",
      "C898\n",
      "C904\n",
      "C906\n",
      "C911\n",
      "C913\n",
      "C914\n",
      "C915\n",
      "C924\n",
      "C925\n",
      "C926\n",
      "C930\n",
      "C931\n",
      "C933\n",
      "C935\n",
      "C937\n",
      "C878\n",
      "C880\n",
      "C882\n",
      "C885\n",
      "C887\n",
      "C888\n",
      "C889\n",
      "C890\n",
      "C892\n",
      "C895\n",
      "C883\n",
      "C879\n",
      "C932\n",
      "C901\n",
      "C902\n",
      "C907\n",
      "C909\n",
      "C884\n",
      "C893\n",
      "C894\n",
      "C896\n",
      "C929\n",
      "C899\n",
      "C934\n",
      "C903\n",
      "C881\n",
      "C916\n",
      "C919\n",
      "C921\n",
      "C920\n",
      "C905\n",
      "C922\n",
      "C910\n",
      "C923\n",
      "C886\n",
      "C912\n",
      "C897\n",
      "C891\n",
      "C908\n",
      "C900\n",
      "C928\n",
      "C917\n",
      "C918\n",
      "C927\n",
      "C936\n"
     ]
    }
   ],
   "source": [
    "# compute same metrics for human loci\n",
    "num_human_loci=[]\n",
    "num_human_seeds = []\n",
    "for focal_clust in clust_df.index.tolist():\n",
    "    print(focal_clust)\n",
    "\n",
    "    focal_genes = clust_df['CD_MemberList'].loc[focal_clust].split(' ')\n",
    "    \n",
    "    if len(np.intersect1d(focal_genes,h_bmi_genes))>0: # only compute loci if there are any seed genes in the cluster\n",
    "        human_bmi_seed_focal=human_bmi_seed_df[human_bmi_seed_df.index.isin(focal_genes)].dropna(subset=['gene_symbol'])\n",
    "        num_human_seeds.append(len(human_bmi_seed_focal))\n",
    "\n",
    "        # how many seed genes are within 250 KB of each other?\n",
    "        start_list = []\n",
    "        end_list = []\n",
    "        chr_list = []\n",
    "        for g in human_bmi_seed_focal.index.tolist():\n",
    "            start_list.append(int(human_bmi_seed_focal.loc[g]['start']))\n",
    "            end_list.append(int(human_bmi_seed_focal.loc[g]['end']))\n",
    "            chr_list.append(human_bmi_seed_focal.loc[g]['chromosome'])\n",
    "\n",
    "        df_temp = pd.DataFrame({'start':start_list,'end':end_list,'chr':chr_list},index=human_bmi_seed_focal.index.tolist())\n",
    "\n",
    "\n",
    "        # loop over all gene pairs, compute distance between start and end, when on same chromosome\n",
    "        g1list = []\n",
    "        g2list = []\n",
    "        dist_list = []\n",
    "        chrom1list = []\n",
    "        chrom2list = []\n",
    "\n",
    "        genes_temp = df_temp.index.tolist()\n",
    "        for i in np.arange(len(genes_temp)-1):\n",
    "            g1 = genes_temp[i]\n",
    "            chr1 = df_temp.loc[g1]['chr']\n",
    "            start1 = df_temp.loc[g1]['start']\n",
    "            end1 = df_temp.loc[g1]['end']\n",
    "            for j in np.arange(i+1,len(genes_temp)): \n",
    "                g2 = genes_temp[j]\n",
    "                chr2 = df_temp.loc[g2]['chr']\n",
    "                start2 = df_temp.loc[g2]['start']\n",
    "                end2 = df_temp.loc[g2]['end']\n",
    "\n",
    "                if chr1==chr2:\n",
    "                    # check if they overlap\n",
    "                    if (start2<end1)&(start1<start2):\n",
    "                        dist_list.append(0)\n",
    "                    elif (start1<end2)&(start2<start1):\n",
    "                        dist_list.append(0)\n",
    "                    elif (start2-end1)>0: # check which is ahead\n",
    "                        dist_list.append(start2-end1)\n",
    "                    else:\n",
    "                        dist_list.append(start1-end2)\n",
    "                else:\n",
    "                    dist_list.append(np.nan) # nans if they are diff chrom\n",
    "\n",
    "                g1list.append(g1)\n",
    "                g2list.append(g2)\n",
    "                chrom1list.append(chr1)\n",
    "                chrom2list.append(chr2)\n",
    "\n",
    "        dist_df_temp = pd.DataFrame({'g1':g1list,'g2':g2list,\n",
    "                                     'chr1':chrom1list,'chr2':chrom2list,\n",
    "                                     'dist':dist_list})\n",
    "\n",
    "\n",
    "\n",
    "        dist_df_temp['within_250KB']=dist_df_temp['dist']<250000\n",
    "        elist_temp = zip(dist_df_temp[dist_df_temp['within_250KB']==True]['g1'],dist_df_temp[dist_df_temp['within_250KB']==True]['g2'])\n",
    "\n",
    "        # make a network from this\n",
    "        G_dist_temp = nx.from_edgelist(elist_temp)\n",
    "\n",
    "        # number communities + number genes not in a shared locus ==> total number of loci\n",
    "\n",
    "        num_human_loci.append(len(pd.Series(community.best_partition(G_dist_temp), dtype=\"object\").value_counts())+len(np.setdiff1d(genes_temp,G_dist_temp.nodes())))\n",
    "    else:\n",
    "        num_human_loci.append(0)\n",
    "        num_human_seeds.append(0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compile everything into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T19:58:37.313665Z",
     "start_time": "2022-07-18T19:58:37.305271Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_human_seeds</th>\n",
       "      <th>num_human_loci</th>\n",
       "      <th>num_rat_seeds</th>\n",
       "      <th>num_rat_loci</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C877</th>\n",
       "      <td>227</td>\n",
       "      <td>162</td>\n",
       "      <td>114</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C898</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C904</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C906</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C911</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      num_human_seeds  num_human_loci  num_rat_seeds  num_rat_loci\n",
       "C877              227             162            114            55\n",
       "C898                4               4              2             2\n",
       "C904                3               3              2             2\n",
       "C906                3               3              1             1\n",
       "C911                4               3              1             1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_loci_df = pd.DataFrame({'num_human_seeds':num_human_seeds,'num_human_loci':num_human_loci,\n",
    "                             'num_rat_seeds':num_rat_seeds,'num_rat_loci':num_rat_loci},index=clust_df.index.tolist())\n",
    "clust_loci_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:00:29.116463Z",
     "start_time": "2022-07-18T20:00:29.110652Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_human_seeds</th>\n",
       "      <th>num_human_loci</th>\n",
       "      <th>num_rat_seeds</th>\n",
       "      <th>num_rat_loci</th>\n",
       "      <th>num_genes_in_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C877</th>\n",
       "      <td>227</td>\n",
       "      <td>162</td>\n",
       "      <td>114</td>\n",
       "      <td>55</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C898</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C904</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C906</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C911</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      num_human_seeds  num_human_loci  num_rat_seeds  num_rat_loci  \\\n",
       "C877              227             162            114            55   \n",
       "C898                4               4              2             2   \n",
       "C904                3               3              2             2   \n",
       "C906                3               3              1             1   \n",
       "C911                4               3              1             1   \n",
       "\n",
       "      num_genes_in_cluster  \n",
       "C877                   642  \n",
       "C898                    11  \n",
       "C904                    10  \n",
       "C906                     9  \n",
       "C911                     7  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_loci_df['num_genes_in_cluster']=clust_df['CD_MemberList_Size']\n",
    "clust_loci_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:00:36.061936Z",
     "start_time": "2022-07-18T20:00:36.051060Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_human_seeds</th>\n",
       "      <th>num_human_loci</th>\n",
       "      <th>num_rat_seeds</th>\n",
       "      <th>num_rat_loci</th>\n",
       "      <th>num_genes_in_cluster</th>\n",
       "      <th>human_loci_per_seed</th>\n",
       "      <th>rat_loci_per_seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C877</th>\n",
       "      <td>227</td>\n",
       "      <td>162</td>\n",
       "      <td>114</td>\n",
       "      <td>55</td>\n",
       "      <td>642</td>\n",
       "      <td>0.713656</td>\n",
       "      <td>0.482456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C878</th>\n",
       "      <td>71</td>\n",
       "      <td>67</td>\n",
       "      <td>36</td>\n",
       "      <td>31</td>\n",
       "      <td>214</td>\n",
       "      <td>0.943662</td>\n",
       "      <td>0.861111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C879</th>\n",
       "      <td>57</td>\n",
       "      <td>55</td>\n",
       "      <td>31</td>\n",
       "      <td>27</td>\n",
       "      <td>169</td>\n",
       "      <td>0.964912</td>\n",
       "      <td>0.870968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C880</th>\n",
       "      <td>60</td>\n",
       "      <td>49</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "      <td>157</td>\n",
       "      <td>0.816667</td>\n",
       "      <td>0.518519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C881</th>\n",
       "      <td>31</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>68</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>0.583333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C882</th>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>53</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C883</th>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C884</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C885</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>31</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C886</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>27</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C887</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C888</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C889</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C891</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C890</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C892</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C893</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C898</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C894</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C896</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C899</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C897</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C895</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C901</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C903</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      num_human_seeds  num_human_loci  num_rat_seeds  num_rat_loci  \\\n",
       "C877              227             162            114            55   \n",
       "C878               71              67             36            31   \n",
       "C879               57              55             31            27   \n",
       "C880               60              49             27            14   \n",
       "C881               31              27             12             7   \n",
       "C882               16              15             11            10   \n",
       "C883                9               8              3             3   \n",
       "C884               10              10              5             5   \n",
       "C885               11              11              6             6   \n",
       "C886                8               8              6             5   \n",
       "C887                3               3              1             1   \n",
       "C888                5               5              2             2   \n",
       "C889                7               7              4             4   \n",
       "C891                5               5              3             3   \n",
       "C890                6               6              1             1   \n",
       "C892                5               5              4             4   \n",
       "C893                5               5              3             3   \n",
       "C898                4               4              2             2   \n",
       "C894                7               7              1             1   \n",
       "C896                1               1              1             1   \n",
       "C899                4               4              3             3   \n",
       "C897                3               3              1             1   \n",
       "C895                3               3              1             1   \n",
       "C901                4               4              3             3   \n",
       "C903                3               3              1             1   \n",
       "\n",
       "      num_genes_in_cluster  human_loci_per_seed  rat_loci_per_seed  \n",
       "C877                   642             0.713656           0.482456  \n",
       "C878                   214             0.943662           0.861111  \n",
       "C879                   169             0.964912           0.870968  \n",
       "C880                   157             0.816667           0.518519  \n",
       "C881                    68             0.870968           0.583333  \n",
       "C882                    53             0.937500           0.909091  \n",
       "C883                    34             0.888889           1.000000  \n",
       "C884                    32             1.000000           1.000000  \n",
       "C885                    31             1.000000           1.000000  \n",
       "C886                    27             1.000000           0.833333  \n",
       "C887                    17             1.000000           1.000000  \n",
       "C888                    17             1.000000           1.000000  \n",
       "C889                    16             1.000000           1.000000  \n",
       "C891                    15             1.000000           1.000000  \n",
       "C890                    15             1.000000           1.000000  \n",
       "C892                    14             1.000000           1.000000  \n",
       "C893                    14             1.000000           1.000000  \n",
       "C898                    11             1.000000           1.000000  \n",
       "C894                    11             1.000000           1.000000  \n",
       "C896                    11             1.000000           1.000000  \n",
       "C899                    11             1.000000           1.000000  \n",
       "C897                    11             1.000000           1.000000  \n",
       "C895                    11             1.000000           1.000000  \n",
       "C901                    10             1.000000           1.000000  \n",
       "C903                    10             1.000000           1.000000  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_loci_df['human_loci_per_seed']=clust_loci_df['num_human_loci']/clust_loci_df['num_human_seeds']\n",
    "clust_loci_df['rat_loci_per_seed']=clust_loci_df['num_rat_loci']/clust_loci_df['num_rat_seeds']\n",
    "clust_loci_df.sort_values('num_genes_in_cluster',ascending=False).head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:04:06.925657Z",
     "start_time": "2022-07-18T20:04:06.912784Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>community</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C877</th>\n",
       "      <td>Conserved BMI Network (root)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C878</th>\n",
       "      <td>Nervous System Development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C879</th>\n",
       "      <td>Synaptic Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C880</th>\n",
       "      <td>Chromosome Organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C881</th>\n",
       "      <td>mRNA Metabolic Process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             annotation\n",
       "community                              \n",
       "C877       Conserved BMI Network (root)\n",
       "C878         Nervous System Development\n",
       "C879                 Synaptic Signaling\n",
       "C880            Chromosome Organization\n",
       "C881             mRNA Metabolic Process"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate with the community names\n",
    "annotations = pd.read_csv(DATADIR + \"outputs/BMI_hierarchy_curated_annotations.tsv\", sep=\"\\t\", index_col=1)\n",
    "annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-18T20:05:56.176462Z",
     "start_time": "2022-07-18T20:05:56.168484Z"
    }
   },
   "outputs": [],
   "source": [
    "# save the table\n",
    "clust_loci_df.join(annotations).sort_index().to_csv(DATADIR + 'outputs/loci_per_community.tsv', sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:rat_pub]",
   "language": "python",
   "name": "conda-env-rat_pub-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
